syncer/streamer_controller.go

```diff
@@ -0,0 +1,343 @@
+					event.Event = ev
+				}
+			}
+		default:
+		}
+		return event, nil
+	}
+
+	if err != context.Canceled {
+		c.meetError = true
+	}
+
+	return nil, err
+}
+
+// ReopenWithRetry reopens streamer with retry
+func (c *StreamerController) ReopenWithRetry(tctx tcontext.Context, pos mysql.Position) error {
+	var err error
+	for i := 0; i < maxRetryCount; i++ {
+		err = c.ResetReplicationSyncer(tctx, pos)
+		if err == nil {
+			return nil
+		}
+		if needRetryReplicate(err) {
+			tctx.L().Info("fail to retry open binlog streamer", log.ShortError(err))
+			time.Sleep(retryTimeout)
+			continue
+		}
+		break
+	}
+	return err
+}
+
+func (c *StreamerController) closeBinlogSyncer(logtctx tcontext.Context, binlogSyncer *replication.BinlogSyncer) error {
+	if binlogSyncer == nil {
+		return nil
+	}
+
+	lastSlaveConnectionID := binlogSyncer.LastConnectionID()
+	defer binlogSyncer.Close()
+	if lastSlaveConnectionID > 0 {
+		err := c.fromDB.killConn(lastSlaveConnectionID)
+		if err != nil {
+			logtctx.L().Error("fail to kill last connection", zap.Uint32("connection ID", lastSlaveConnectionID), log.ShortError(err))
+			if !utils.IsNoSuchThreadError(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Close closes streamer
+func (c *StreamerController) Close(tctx tcontext.Context) {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.streamerProducer != nil {
+		switch r := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			// process remote binlog reader
+			c.closeBinlogSyncer(tctx, r.reader)
+			c.streamerProducer = nil
+		case *localBinlogReader:
+			// process local binlog reader
+			r.reader.Close()
+		}
+	}
+}
+
+func (c *StreamerController) setUUIDIfExists(tctx tcontext.Context, filename string) bool {
+	_, uuidSuffix, _, err := binlog.AnalyzeFilenameWithUUIDSuffix(filename)
+	if err != nil {
+		// don't contain uuid in position's name
+		return false
+	}
+
+	c.uuidSuffix = uuidSuffix
+	return true
+}
+
+// UpdateSyncCfg updates sync config and fromDB
+func (c *StreamerController) UpdateSyncCfg(syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn) {
+	c.Lock()
+	c.fromDB = fromDB
+	c.syncCfg = syncCfg
+	c.Unlock()
+}
+
+// check whether the uuid in binlog position's name is same with upstream
+func (c *StreamerController) checkUUIDSameWithUpstream(pos mysql.Position, uuids []string) (bool, error) {
+	_, uuidSuffix, _, err := binlog.AnalyzeFilenameWithUUIDSuffix(pos.Name)
+	if err != nil {
+		// don't contain uuid in position's name
+		return true, nil
+	}
+	uuid := utils.GetUUIDBySuffix(uuids, uuidSuffix)
+
+	upstreamUUID, err := utils.GetServerUUID(c.fromDB.BaseDB.DB, c.syncCfg.Flavor)
+	if err != nil {
+		return false, terror.Annotate(err, "streamer controller check upstream uuid failed")
+	}
+
+	return uuid == upstreamUUID, nil
+}
+
+// GetBinlogType returns the binlog type used now
+func (c *StreamerController) GetBinlogType() BinlogType {
+	if c.streamerProducer == nil {
+		return c.binlogType
+	}
+
+	switch c.streamerProducer.(type) {
```

Comment:
How about updating `c.binlogType` every time we change `s.streamerProducer` or change `binlogType` to `initialBinlogType`? It might be confusing that `binlogType` is not current binlog type.

Reply:
refine in aebdabe